watertap/unit_models/Bipolar_Electrodialysis_0D.py

```diff
@@ -0,0 +1,2343 @@
+        self.acidic = ControlVolume0DBlock(
+            dynamic=False,
+            has_holdup=False,
+            property_package=self.config.property_package,
+            property_package_args=self.config.property_package_args,
+        )
+        self.acidic.add_state_blocks(has_phase_equilibrium=False)
+        self.acidic.add_material_balances(
+            balance_type=self.config.material_balance_type, has_mass_transfer=True
+        )
+        self.acidic.add_energy_balances(
+            balance_type=self.config.energy_balance_type,
+            has_enthalpy_transfer=False,
+        )
+
+        if self.config.is_isothermal:
+            self.acidic.add_isothermal_assumption()
+        self.acidic.add_momentum_balances(
+            balance_type=self.config.momentum_balance_type,
+            has_pressure_change=self.config.has_pressure_change,
+        )
+
+        # den_mass and visc_d in acidic and basic channels are the same
+        add_object_reference(
+            self, "dens_mass", self.acidic.properties_in[0].dens_mass_phase["Liq"]
+        )
+        add_object_reference(
+            self, "visc_d", self.acidic.properties_in[0].visc_d_phase["Liq"]
+        )
+
+        # Add ports (creates inlets and outlets for each channel)
+        self.add_inlet_port(name="inlet_basic", block=self.basic)
+        self.add_outlet_port(name="outlet_basic", block=self.basic)
+        self.add_inlet_port(name="inlet_acidic", block=self.acidic)
+        self.add_outlet_port(name="outlet_acidic", block=self.acidic)
+
+        # extension options
+        if self.config.has_catalyst == True:
+            self._make_catalyst()
+
+        if (
+            not self.config.pressure_drop_method == PressureDropMethod.none
+        ) and self.config.has_pressure_change:
+            self._pressure_drop_calculation()
+
+            @self.Constraint(
+                self.flowsheet().time,
+                doc="Pressure drop expression as calculated by the pressure drop data, "
+                "base channel of the bipolar membrane.",
+            )
+            def eq_deltaP_basic(self, t):
+                return self.basic.deltaP[t] == -self.pressure_drop_total[t]
+
+            @self.Constraint(
+                self.flowsheet().time,
+                doc="Pressure drop expression as calculated by the pressure drop data,"
+                "  acid channel of the bipolar membrane.",
+            )
+            def eq_deltaP_acidic(self, t):
+                return self.acidic.deltaP[t] == -self.pressure_drop_total[t]
+
+        elif self.config.pressure_drop_method == PressureDropMethod.none and (
+            not self.config.has_pressure_change
+        ):
+            pass
+        else:
+            raise ConfigurationError(
+                "A valid (not none) pressure_drop_method and has_pressure_change being True "
+                "must be both used or unused at the same time. "
+            )
+
+        # Build Constraints
```

Comment:
Note for myself, only got to here on review but wanted to publish initial comments